pkg/istio-agent/xds-agent.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istioagent
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"time"
+
+	"google.golang.org/grpc"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/proxy/envoy/xds"
+	"istio.io/istio/pkg/adsc"
+	"istio.io/pkg/env"
+	"istio.io/pkg/log"
+)
+
+// xds-agent runs an XDS server in agent, similar with the SDS server. It is using the same
+// UDS socket as the SDS server - envoy will use a single connection/cluster for both.
+//
+// Envoy will connect to the agent using UDS. Agent will handle connection to one
+// or more upstream XDS servers, and return aggregated and possibly patched results.
+
+// This is part of an optimization experiment and exploring federation and local
+// transformations - not enabled by default and not intended for production use yet.
+
+var (
+	udsListener net.Listener
+
+	// Used for debugging. Will also be used to recover in case XDS is unavailable, and to evaluate the
+	// caching. Key format and content should follow the new resource naming spec, will evolve with the spec.
+	// As Envoy implements the spec, they should be used directly by Envoy - while prototyping they can
+	// be handled by the proxy.
+	savePath = env.RegisterStringVar("XDS_SAVE", "./var/istio/xds",
+		"If set, the XDS proxy will save a snapshot of the received config")
+)
+
+var (
+	// Can be used by Envoy or istioctl debug tools.
+	xdsAddr = env.RegisterStringVar("XDS_LOCAL", "127.0.0.1:15002",
```

Comment:
- If _istioctl_ uses this, would I change it to 0.0.0.0:15002?
- Does this makes the debug API visible to the app container by default?  This seems bad for security.  Perhaps the default should be `""` meaning don't offer XDS local proxy.

Reply:
definitely should not be 0.0.0.0, that is extremely insecure. re: app container, the envoy config dump is already visible

Reply:
Yes, no different from config dump - we can expose both over UDS if you can exec to the container.

We should never expose plain text over 0.0.0.0 ( i.e. without authn/authz).